### PR TITLE
Themes: Bootstrap theme-details store for theme sheets

### PR DESCRIPTION
**Pull request by seear**
* Add the theme-details data component to the layout
* Fetch and cache theme details server-side and bootstrap to the client

screen shot 2016-02-23 at 21 24 01

**To Test**
Go to http://calypso.localhost:3000/themes/{theme} both logged-in and logged-out

### CODE REVIEW

File: webpack.config.js [hunk 1]
```diff
@@ -97,7 +97,7 @@ jsLoader = {
 };
 
 if ( CALYPSO_ENV === 'development' ) {
-	webpackConfig.plugins.push( new PragmaCheckPlugin() );
+//	webpackConfig.plugins.push( new PragmaCheckPlugin() );
```
Comment (seear) [inline webpack.config.js]: Problem: `state/themes/actions.js` is replaced with noop in `webpack.config.node.js` but the pragma checker doesn't know this, so we still get the `not /** @ssr-ready **/` error. @ehg: any idea how we should handle this?

Comment (ehg) [inline webpack.config.js]: Hmm, we can add analytics to the ignore array, but I think `state/themes/actions.js` should be ssr-ready from the start. Will take a look.

Comment (ehg) [inline webpack.config.js]: Pushed acc169f

File: server/pages/index.js [hunk 2]
```diff
@@ -376,20 +378,44 @@ module.exports = function() {
...
 				) ) );
-			}
+				res.render( 'index.jade', context );
+			};
 
-			res.render( 'index.jade', context );
+			if ( config.isEnabled( 'server-side-rendering' ) ) {
+				let themeData = themeDetails.get( req.params.theme_slug );
+				if ( ! themeData ) {
+					wpcom.undocumented().themeDetails( req.params.theme_slug, ( error, data ) => {
```
Comment (ehg) [inline server/pages/index.js]: I wonder if we could get away without having to use memcache by:
- Checking the published date of the theme details response
- If it's greater than the one in the cache, `renderToString` using the new data.
- If it's not, just serve the cached `renderToString` markup

We could also avoid blocking the request by always serving the data from the cache, and fire off API calls that will invalidate/update the cache when necessary.

**Comment by ehg** [comment]
Rebased, not using `makeElement()` for now.

File: server/pages/index.js (hunk 2 above)
Comment (ockham) [inline server/pages/index.js]: I like that idea :-)

**Comment by ehg** [comment]
Changed to 'In Progress', as this should be based off #3386, and not merged until that lands. Also, we need more discussion around  https://github.com/Automattic/wp-calypso/pull/3279#discussion_r53366518

File: server/pages/index.js (hunk 2 above)
Comment (seear) [inline server/pages/index.js]: > We could also avoid blocking the request by always serving the data from the cache, and fire off API calls that will invalidate/update the cache when necessary.

This PR was actually working like that a few commits back and can easily be tweaked to do that again. I wasn't quite sure which way was best. It strikes me now that never rendering async is less complex overall.

**Comment by ehg** [comment]
> this should be based off #3386

Done.

File: server/pages/index.js (hunk 2 above)
Comment (seear) [inline server/pages/index.js]: > I wonder if we could get away without having to use memcache by...

Yeah, was thinking about this, or maybe even using the api response header, which may let us leverage some existing libraries to handle it?

Once we receive new api data, how do we clear the renderToString cache?

Comment (ehg) [inline server/pages/index.js]: > Yeah, was thinking about this, or maybe even using the api response header, which may let us leverage some existing libraries to handle it?

Yep. I wonder whether we sent ETags or what the Expires header is like.

> Once we receive new api data, how do we clear the renderToString cache?

You mean how do we know what key to clear? Maybe we key the rts cache using something like `dataTimestamp+lang`?

Comment (ehg) [inline server/pages/index.js]: > I wonder whether we sent ETags or what the Expires header is like.

Expires and cache control headers are explicitly set to not cache, which makes sense. I think comparing the publish date is probably fine for now.

File: server/pages/index.js [hunk 3]
```diff
@@ -375,20 +378,44 @@ module.exports = function() {
...
-			if ( config.isEnabled( 'server-side-rendering' ) ) {
+			const renderThemeSheet = function( theme ) {
 				const store = createReduxStore();
+				store.dispatch( {
+					type: ActionTypes.RECEIVE_THEME_DETAILS,
+					themeId: theme.id,
+					themeName: theme.name,
+					themeAuthor: theme.author,
+					themeScreenshot: theme.screenshot,
+				} );
```
Comment (ockham) [inline server/pages/index.js]: Could this call go somewhere deeper in the component hierarchy, e.g. inside `ThemesListFetcher` (to provide some initial state there, also on the client side)? Thinking about how to keep this part of `server/pages` as generic as possible for isomorphic routing.

File: server/pages/index.js [hunk 4]
```diff
@@ -18,7 +18,9 @@ var config = require( 'config' ),
 	LayoutLoggedOutDesign = require( 'layout/logged-out-design' ),
 	render = require( 'render' ).render,
 	createReduxStore = require( 'state' ).createReduxStore,
-	setSection = require( 'state/ui/actions' ).setSection;
+	setSection = require( 'state/ui/actions' ).setSection,
+	wpcom = require( 'lib/wp' ),
+	ActionTypes = require( 'state/themes/action-types' );
```
Comment (mcsf) [inline server/pages/index.js]: Should be `ThemesActionTypes`

File: server/pages/index.js (hunk 3 above)
Comment (mcsf) [inline server/pages/index.js]: Actually, `renderThemeSheet` looks like something we could keep as middleware in `my-sites/themes/controller`: it basically writes to `context` and dispatches stuff via `store`, which should be available via `context`.

**Comment by ehg** [comment]
Rebased.

**Comment by ehg** [comment]
Added dynamic data to content sections and fixed price. Next up: make code not suck.

File: server/pages/index.js (hunk 2 above)
Comment (ehg) [inline server/pages/index.js]: RE https://github.com/Automattic/wp-calypso/pull/3279#discussion_r53366518 - implemented a basic version of this in b560f0a

**Comment by seear** [comment]
rebased after #3464

File: client/my-sites/themes/controller.js [hunk 5]
```diff
@@ -130,7 +131,16 @@ export function details( context, next ) {
 		isFullScreen: true
 	} ) );
 
-	context.primary = makeElement( ThemeSheetComponent, Head, context.store, props );
 	context.secondary = null; // When we're logged in, we need to remove the sidebar.
+	//TODO: use makeElement()
+	context.primary = (
+		<ReduxProvider store={ context.store } >
+			<Head title={ props.title } isSheet>
```
Comment (seear) [inline client/my-sites/themes/controller.js]: We're still using the hardcoded _Pineapple Fifteen_ title, but I think that should be hooked up in a separate PR https://github.com/Automattic/wp-calypso/blob/339c6bbb5a50d975e96c1c2c016b498044fc9107/client/my-sites/themes/head.jsx#L15

File: server/pages/index.js [hunk 6]
```diff
@@ -374,22 +377,57 @@ module.exports = function() {
 	} );
 
 	if ( config.isEnabled( 'manage/themes/details' ) ) {
+		function updateRenderCache( themeSlug ) {
+			let themeData = themeDetails.get( themeSlug );
+			wpcom.undocumented().themeDetails( themeSlug, ( error, data ) => {
+				if ( error ) {
+					console.log( `Error fetching theme ${ themeSlug } details: `, error.message || error );
```
Comment (seear) [inline server/pages/index.js]: I've added a todo item to #2875 for a 404/Drake when theme not found

File: server/pages/index.js [hunk 7]
```diff
@@ -374,22 +377,57 @@ module.exports = function() {
...
+					console.log( 'found theme!', theme.id );
+				}
 
-				Object.assign( context, render( (
-					<ReduxProvider store={ store }>
-						<LayoutLoggedOutDesign store={ store } routeName={ 'themes' } match={ { theme_slug: req.params.theme_slug } } />
-					</ReduxProvider>
-				) ) );
+				i18n.initialize();
+				req.params.theme_slug && updateRenderCache( req.params.theme_slug ); // TODO(ehg): We don't want to hit the endpoint for every req. Debounce based on theme arg?
```
Comment (seear) [inline server/pages/index.js]: > We don't want to hit the endpoint for every req.

Why not?

File: server/pages/index.js [hunk 8]
```diff
@@ -374,22 +377,57 @@ module.exports = function() {
...
 	if ( config.isEnabled( 'manage/themes/details' ) ) {
+		function updateRenderCache( themeSlug ) {
+			let themeData = themeDetails.get( themeSlug );
+			wpcom.undocumented().themeDetails( themeSlug, ( error, data ) => {
+				if ( error ) {
+					console.log( `Error fetching theme ${ themeSlug } details: `, error.message || error );
+				} else if ( ! themeDetails.has( themeSlug ) || ( Date( data.date_updated ) > Date( themeData.date_updated ) ) ) {
+					console.log( 'caching', themeSlug );
+					themeDetails.set( themeSlug, data );
+					renderThemeSheet( data );
```
Comment (seear) [inline server/pages/index.js]: I'm leaning towards not doing this render. The less stuff that happens async, the simpler it is. For example, we would have to re-init i18n with the appropriate locale (once we support other locales).

Comment (seear) [inline server/pages/index.js]: Oh, I see, this just updates the renderToString cache. Got it.

File: server/pages/index.js (hunk 7 above)
Comment (ehg) [inline server/pages/index.js]: More events in the loop? Maybe it's a pre-optimisation, although I wonder if we rate limit the API on the internal network.

File: server/pages/index.js (hunk 6 above)
Comment (ehg) [inline server/pages/index.js]: Cool. I've added one for loading placeholders.

File: client/my-sites/themes/controller.js (hunk 5 above)
Comment (ehg) [inline client/my-sites/themes/controller.js]: Yeah, I think we should leave this until after routing. It's possibly being implemented in one of those PRs.

File: server/pages/index.js [hunk 9]
```diff
@@ -374,22 +377,57 @@ module.exports = function() {
...
+				themePrice: theme.price ? theme.price.display : undefined,
+				themeDescription: theme.description,
+				themeDescriptionLong: theme.description_long,
+				themeSupportDocumentation: theme.extended ? theme.extended.support_documentation : undefined,
+			} );
+
+			store.dispatch( setSection( 'themes', { hasSidebar: false, isFullScreen: true } ) );
+			context.initialReduxState = pick( store.getState(), 'ui', 'themes' );
+
+			Object.assign( context, render( (
```
Comment (seear) [inline server/pages/index.js]: do we need to pass `date_updated` here to force another `renderToString`?

File: server/pages/index.js (hunk 8 above)
Comment (ehg) [inline server/pages/index.js]: Yep. It's kind of confusing, hence my comment about using a more explicit cache.

File: client/my-sites/themes/controller.js (hunk 5 above)
Comment (ockham) [inline client/my-sites/themes/controller.js]: Maybe this stuff, with some modifications? -> https://github.com/Automattic/wp-calypso/pull/3425/files#diff-85f90b61d9c36634c5464a3180470726R70

File: server/render/index.js [hunk 10]
```diff
@@ -23,18 +21,18 @@ function bumpStat( group, name ) {
...
 }
 
-export function render( element ) {
-	memoizedRenderToString.cache = markupCache;
-
-	if ( ! memoizedRenderToString.cache.has( JSON.stringify( element ) ) ) {
-		bumpStat( 'calypso-ssr', 'loggedout-design-cache-miss' );
-	}
-
+export function render( element, path, initialState ) {
```
Comment (seear) [inline server/render/index.js]: Maybe better to have a `key` param instead of `path` and `initialState`, allowing caller to decide appropriate key?

Comment (ehg) [inline server/render/index.js]: Yeah, also I was thinking the API could be something like `render.set( key, element )` and `render.get( key )`

Comment (seear) [inline server/render/index.js]: > Yeah, also I was thinking the API could be something like render.set( key, element ) and render.get( key )

I think better to put a comment before the async `renderThemeSheet()` than expose the caching from the render module.

File: server/pages/index.js [hunk 11]
```diff
@@ -375,23 +380,64 @@ module.exports = function() {
 	} );
 
 	if ( config.isEnabled( 'manage/themes/details' ) ) {
-		app.get( '/themes/:theme_slug', function( req, res ) {
-			const context = getDefaultContext( req );
+		app.get( '/themes/:theme_slug/:section?', function( req, res ) {
+			function updateRenderCache( themeSlug ) {
+				wpcom.undocumented().themeDetails( themeSlug, ( error, data ) => {
+					if ( error ) {
+						console.log( `Error fetching theme ${ themeSlug } details: `, error.message || error );
```
Comment (ehg) [inline server/pages/index.js]: Can we use `debug` for these console.logs?

**Comment by ehg** [comment]
:+1: after squash. Let's get this in and iterate.

File: server/pages/index.js [hunk 12]
```diff
@@ -375,23 +380,64 @@ module.exports = function() {
...
+				const key = JSON.stringify( element ) + path + JSON.stringify( context.initialReduxState );
+				Object.assign( context, render( element, key ) );
+				return context;
+			};
+
+			const context = getDefaultContext( req );
+			if ( config.isEnabled( 'server-side-rendering' ) ) {
+				const theme = themeDetails.get( req.params.theme_slug );
+				if ( theme ) {
+					Object.assign( context, renderThemeSheet( theme ) );
```
Comment (ockham) [inline server/pages/index.js]: Is it possible that this code path ends up renderingToString synchronously via `renderThemeSheet()` and `render()`, i.e. blocking the request?

**Comment by ockham** [comment]
I find the two different caches and how they interact rather confusing, hard to reconcile with #3562, and possibly error-prone. If we want to cache the markup based on `date_updated`, can't we just make it part of the markup cache key [here](https://github.com/Automattic/wp-calypso/pull/3279/files#diff-b2d6784517d1151d7e75f0287d0c6478R426) -- possibly by just exposing it as part of the `themeDetails` object we dispatch [here](https://github.com/Automattic/wp-calypso/pull/3279/files#diff-b2d6784517d1151d7e75f0287d0c6478R403), so it becomes part of the [`initialReduxState`](https://github.com/Automattic/wp-calypso/pull/3279/files#diff-b2d6784517d1151d7e75f0287d0c6478R416), which is then used for the key?

If we just want to cache the theme details data for their own sake, we should do that inside our Redux reducer -- it's made for storing data for different theme IDs already, so we're kind of re-creating that logic here.

File: server/pages/index.js [hunk 13]
```diff
@@ -375,23 +380,64 @@ module.exports = function() {
...
+					if ( error ) {
+						debug( `Error fetching theme ${ themeSlug } details: `, error.message || error );
+						return;
+					}
+					const themeData = themeDetails.get( themeSlug );
+					if ( ! themeData || ( Date( data.date_updated ) > Date( themeData.date_updated ) ) ) {
+						debug( 'caching', themeSlug );
+						themeDetails.set( themeSlug, data );
+						// update the render cache
+						renderThemeSheet( data );
```
Comment (ockham) [inline server/pages/index.js]: Why don't we also render `index.jade` here?

File: server/pages/index.js [hunk 14]
```diff
@@ -375,23 +380,64 @@ module.exports = function() {
...
 
-				Object.assign( context, render( (
+				const element = (
 					<ReduxProvider store={ store }>
 						<LayoutLoggedOut primary={ primary } />
 					</ReduxProvider>
-				) ) );
+				);
+				const path = url.parse( req.url ).path;
+				const key = JSON.stringify( element ) + path + JSON.stringify( context.initialReduxState );
```
Comment (ockham) [inline server/pages/index.js]: Wouldn't we ideally use the the props that `element` receives (possibly including `date_updated`, as [suggested](https://github.com/Automattic/wp-calypso/pull/3279#issuecomment-191976564)) as key? Which would be congruent with the relevant information for `shouldComponentUpdate`, I think.

**Comment by ockham** [comment]
Filed #3774 based on my comments.
